### PR TITLE
SONARSEC-45: Add target object as first argument to method call in uCFG

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBlockBuilder.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ public BasicBlock CreateBasicBlock(Block block)
             var ucfgBlock = CreateBlockWithId(blockIdProvider.Get(block));
 
             ucfgBlock.Instructions.AddRange(
-                block.Instructions.Select(instructionBuilder.Create).WhereNotNull());
+                block.Instructions.SelectMany(instructionBuilder.Create).WhereNotNull());
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBlockBuilder.cs]: I think this `WhereNotNull` should not be needed, as Create would return Enumerable.Empty when no instruction was created

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs [hunk 2]
```diff
@@ -85,123 +85,154 @@ public Instruction Create(SyntaxNode syntaxNode)
...
             {
-                CreateMethodCall(attributeSyntax, UcfgMethod.Create(attributeCtor)),
+                AddMethodCall(attributeSyntax, attributeCtor),
                 CreateParameterAnnotation(parameterName, attributeSyntax),
             };
 
-        private Instruction CreateDefaultInstruction(SyntaxNode node) =>
-            CreateConstant(node);
+        private IEnumerable<Instruction> CreateDefaultInstruction(SyntaxNode node) =>
+            new[] { CreateConstant(node) };
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs]: CreateConstant could return Enumerable.Empty and you will not be required to create a new array here and in many other places

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs [hunk 3]
```diff
@@ -85,123 +85,154 @@ public Instruction Create(SyntaxNode syntaxNode)
...
-                }
+                return Enumerable.Empty<Instruction>();
             }
 
-            return null;
+            var variable = semanticModel.GetDeclaredSymbol(variableDeclarator);
+            if (IsLocalVarOrParameter(variable))
+            {
+                return new[] { CreateAssignment(variableDeclarator, variable.Name,
+                    GetMappedExpression((variableDeclarator.Initializer.Value))) };
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs]: double parenthesis here

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs [hunk 4]
```diff
@@ -135,31 +132,23 @@ public class Class1
...
-
             ObjectProperty = s;         // %9 := Namespace.Class1.ObjectProperty.set [ s ]
-
-            ObjectProperty = 5;         // ignored
-
-            var x = IntProperty = 5;    // ignored
-
+            ObjectProperty = 5;         // %10 := Namespace.Class1.ObjectProperty.set [ const ]
+            var x = IntProperty = 5;    // %11 := Namespace.Class1.IntProperty.set [ const ]
+                                        // x := __id [ %11 ]
```
Comment (valhristov) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs]: add tests with `this.Property = s` and `other.Property = s` the correct instance should be passed

Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs]: in general all tests should include cases with `this` `base` (where needed) and `other`

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs [hunk 5]
```diff
@@ -233,32 +215,24 @@ public class Class1
...
+                                            // a := __id [ %5 ]
+            a = A(B(C(s)));         // %6 := Namespace.Class1.C(string) [ this s ]
+                                    // %7 := Namespace.Class1.B(int) [ this %6 ]
+                                    // %8 := Namespace.Class1.A(int) [ this %7 ]
+                                    // a := __id [ %8 ]
             int x;
-            x = O(s);               // %8 := Namespace.Class1.O(object) [ s ];
-                                    // no assignment is added because O returns int
+            x = O(s);               // %9 := Namespace.Class1.O(object) [ this s ]
+                                    // x := __id [ %9 ]
```
Comment (valhristov) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs]: Also tests with `this.A()` and `other.A()` (let me know if you need details on what I mean)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs (hunk 4 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs]: done

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs (hunk 5 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Security/Ucfg/UcfgBuilder_Instructions.cs]: done

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs [hunk 6]
```diff
@@ -80,192 +82,242 @@ public Instruction Create(SyntaxNode syntaxNode)
...
+                {
+                    yield return CreateStaticCallExpression(property.ContainingType);
+                }
+                else
                 {
-                    return CreateAssignment(
-                        variableDeclarator,
-                        variable.Name,
-                        variableDeclarator.Initializer.Value);
+                    yield return ThisExpression;
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs]: what about "var x = other.Property"? isn't it going to return `%0 Property [this]`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs [hunk 7]
```diff
@@ -80,192 +82,242 @@ public Instruction Create(SyntaxNode syntaxNode)
...
+            {
+                if (methodSymbol.IsStatic ||
+                    methodSymbol.ReducedFrom != null)
+                {
+                    yield return CreateStaticCallExpression(methodSymbol.ContainingType);
+                }
+
+                if (!methodSymbol.IsStatic)
+                {
+                    if (invocationExpression.Expression is MemberAccessExpressionSyntax memberAccess)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs]: I think this code should be applied to properties

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs [hunk 8]
```diff
@@ -99,6 +99,11 @@ public static UcfgMethod Create(IMethodSymbol methodSymbol)
             }
         }
 
+        public static UcfgMethod Create(INamedTypeSymbol typeSymbol)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs]: [for later] UcfgMethod should become something like `UcfgTypeHelper` with `GetMethodId` and `GetClassName` or to be split into two...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs (hunk 7 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgInstructionBuilder.cs]: I am not sure to understand what you mean

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs (hunk 8 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgMethod.cs]: Good point... I'll do that now

**Pull request merged**